docs/design/sql/16-nested-fields-access.md

```diff
@@ -0,0 +1,359 @@
+at run-time when that mapping is used.
+
+## Behavior of custom types
+
+All the behavior of `OBJECT` will also apply to UDTs, but additionally:
+
+- one can access fields of the value using the DOT operator (`.`)
+
+This has notable consequences:
+
+- when a custom-typed value is returned to the client, the value will be the
+  java instance or a `GenericRecord`. Non-java clients must use `TO_ROW`
+  function.
+
+- comparison operators (and most notably, the `=` operator) will compare the
+  java instances, not the values field-by-field. Again, one can use `TO_ROW` to
+  compare field-by-field. This might seem counter-intuitive, but it is the
+  limitation already present: if the imap's `__key` field is compared, we compare
+  the instances and not the individual fields, if they are expanded.
+
+## Case sensitivity
+
+Unquoted custom type names, same as all other schema objects, are
+case-sensitive. This is contrary to built-in types, whose names are
+case-insensitive. This might seem illogical, but is consistent with the rest of
+our SQL implementation where identifiers are case-sensitive, but keywords (which
+include built-in functions and types) are not.
+
+```sql
+CREATE TYPE My_Type ...;
+
+CREATE MAPPING m (
+  field my_type  -- fails here
+) ...
+```
+
+Here we need to note that identifiers in SQL are always case-sensitive, however,
+many SQL implementations convert unquoted identifiers to canonical case, which
+gives the impression that identifiers are case-insensitive. Quoted identifiers
+are always case-sensitive. The conversion to canonical case is optional in the
+SQL standard, Hazelcast doesn't do it for legacy reasons.
+
+## Automatic resolution
+
+When a mapping or another type is auto-resolved, for non-primitive types we
+check the SQL catalog if a type exists.
+
+## UDTs and the top-level fields
+
+Currently, if the top-level `__key` or `this` field of an imap isn't a SQL
+primitive type, we make the field available as a hidden field, and all the
+fields are expanded as top-level fields. For example, if we have `IMap<Long,
+Person>`, then the mapping will have these fields:
+
+- `__key BIGINT`
+- `this OBJECT` (hidden)
+- `name`, `birthDate`, ... (all fields of the `Person` class)
+
+This behavior remains unchanged, that is, even if there is a UDT for `Person`,
+`this` will have `OBJECT` type. 
+
+## ROW values
+
+The SQL standards specifies the ROW type. A ROW value contains an array of
+values, but doesn't contain field names. It is important to note that UDT values
+aren't ROW values. However, they can be converted both ways:
+
+- `TO_ROW` function converts an UDT value to `ROW` value
+
+- `CAST(rowExpression AS <UDT name>)`: converts a `ROW` value to a UDT value.
+  The ROW value must have the same number and types of fields as the UDT, in the
+  same order.
+
+When the ROW value is returned to the client, a
+`com.hazelcast.sql.impl.expression.RowValue` is returned to the java client.
```

Comment:
Currently the codec encodes it as String, so in effect RowValue is never returned to the client (HZ-1379). Not sure what would be the ideal solution here, but the current implementation of the codec is effectively a placeholder right now.